### PR TITLE
Added CI pipeline stage descriptions. ST-145

### PR DESCRIPTION
**Pull request by SKAJohanVenter**
Added the Pipeline stage descriptions to the continuous integration page.

JIRA: ST-145

### CODE REVIEW

File: src/tools/continuousintegration.rst [hunk 1]
```diff
@@ -89,6 +90,132 @@ To do that the option must be enabled:
...
+_____________
+- Every commit could potentially trigger a pipeline build. There may be different rules applied to determine which stages are executed in the pipeline based on factors like the branch name.
+
+    - E.g Every commit in a feature branch may trigger the “Lint” stage, but not a slow test suite.
+- When doing a release with a git tag, the full pipeline will be run.
+- Every pipeline job is associated with its git commit (including tag commits).
+- Try and have the stages complete as fast as possible.
+
+    - In some cases it may be possible to parallelize jobs. For example, unit tests and static analysis could be run in parallel.
+- If a job fails, the pipeline will be aborted and the cause of the failure needs to be addressed.
```
Comment (flyingfrog81) [inline src/tools/continuousintegration.rst]: not necessarily true for all stages. Still you might want to execute other stages and collect metrics without publishing the final product.

File: src/tools/continuousintegration.rst [hunk 2]
```diff
@@ -89,6 +90,132 @@ To do that the option must be enabled:
...
+- All projects must include all the stages listed below.
+- Project dependencies must be stored in, and made available from the SKA software repository.
+- All tests must pass on the “master” branch and should be kept stable.
+- Every stage should have some kind of JUnit output to indicate success or failure. In the case of failure it should display the error messages.
```
Comment (flyingfrog81) [inline src/tools/continuousintegration.rst]: not Every stage. This is certainly tru for Test stages but I would remove this bullett point.

File: src/tools/continuousintegration.rst [hunk 3]
```diff
@@ -89,6 +90,132 @@ To do that the option must be enabled:
...
+Stages
+______
+Build
+"""""
+The build stage packages/compiles the software project into distributable units of software.
+The project will be checked out at the git commit hash. This specific version of the code must then be built. Failing the build stage will stop the further steps from being executed. Where possible Semantic Versioning should be used.
+To create a release a git tag should be used. `See Release Procedure <http://developer.skatelescope.org/en/latest/development/python_package_release_procedure.html>`_.
+
+Input
+  Git commit hash (provided by Gitlab)
```
Comment (flyingfrog81) [inline src/tools/continuousintegration.rst]: remove reference to gitlab

File: src/tools/continuousintegration.rst [hunk 4]
```diff
@@ -89,6 +90,132 @@ To do that the option must be enabled:
...
+Input
+  The output from the Build stage. E.g .deb or .whl or docker image.
+  Input could also consist of test data or environment.
+
+Output
+  - The results of the tests in JUnit format. These need to be added to the artifacts.
+    `See Gitlab Test Reports <https://docs.gitlab.com/ee/ci/junit_test_reports.html>`_.
+  - Coverage metrics in JUnit format.
+
+Test types
```
Comment (flyingfrog81) [inline src/tools/continuousintegration.rst:155]: @giobra this is still a work in progress and this is not intended to be a comprehensive description of possible test types. Can you have a look at this? Do you think it makes sense to have it here?

**Review by flyingfrog81 — CHANGES_REQUESTED** [review]
Overall ok, I have asked @giobra to review the testing part

File: src/tools/continuousintegration.rst (hunk 1 above)
Comment (SKAJohanVenter) [reply src/tools/continuousintegration.rst]: 👍

File: src/tools/continuousintegration.rst (hunk 2 above)
Comment (SKAJohanVenter) [reply src/tools/continuousintegration.rst]: 👍

File: src/tools/continuousintegration.rst (hunk 3 above)
Comment (SKAJohanVenter) [reply src/tools/continuousintegration.rst]: 👍

**Review by flyingfrog81 — DISMISSED** [review]

**Comment by flyingfrog81** [comment]
If @giobra can have a look at the testing part quickly, otherwise we shall merge this

File: src/tools/continuousintegration.rst (hunk 4 above)
Comment (giobra) [reply src/tools/continuousintegration.rst:155]: Sorry for jumping in so late. I was not aware of this.

Sounds good. We might add some detail such as "very fast ==> each test takes no more than a handful of ms". 
What is a "component"? a dockerized service/tool?

Why components are not mocked in integration tests? In fact, if you want to really test only the interace between 2 components, then you better mock everything other than the interface code.

Deployment tests: are like smoke tests, aka semanitc monitoring tests? if so some more details would be needed to let programmers understand what this stage is about. 



We might discuss these things in more detail within the tetsing CoP (f2f meeting at Pune?) also in the context of the SKA testing strategy. The testing strategy infact should specify all the relevant test levels/stages and provide definitions and explanations.

Comment (flyingfrog81) [reply src/tools/continuousintegration.rst:155]: thanks @giobra. Good, I guess we will provide the scaffolding for now and the detailed definitions will be filled in with the help of the CoP. 
@SKAJohanVenter can you adress the comments abour "very fast" and maybe add some `..todo` directives where you see further definition will be needed. I still think I'd like to have the page published.

Comment (SKAJohanVenter) [reply src/tools/continuousintegration.rst:155]: Thanks @giobra and @flyingfrog81 
I updated the docs accordingly.

**Review by flyingfrog81 — APPROVED** [review]
LGTM

**Pull request merged**